common/peerprovider/ringpopprovider/provider.go

```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		return nil, fmt.Errorf("ringpop discovery provider: %v", err)
+	}
+
+	bootstrapOpts := &swim.BootstrapOptions{
+		MaxJoinDuration:  config.MaxJoinDuration,
+		DiscoverProvider: discoveryProvider,
+	}
+
+	return NewRingpopProvider(service, rp, bootstrapOpts, logger), nil
+}
+
+// NewRingpopProvider sets up ringpop based peer provider
+func NewRingpopProvider(service string,
+	rp *ringpop.Ringpop,
+	bootstrapOpts *swim.BootstrapOptions,
+	logger log.Logger,
+) *Provider {
+	return &Provider{
+		service:     service,
+		status:      common.DaemonStatusInitialized,
+		ringpop:     rp,
+		bootParams:  bootstrapOpts,
+		logger:      logger,
+		subscribers: map[string]chan<- *membership.ChangedEvent{},
+	}
+}
+
+// Start starts ringpop
+func (r *Provider) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&r.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	_, err := r.ringpop.Bootstrap(r.bootParams)
+	if err != nil {
+		r.logger.Fatal("unable to bootstrap ringpop", tag.Error(err))
+	}
+
+	// Get updates from ringpop ring
+	r.ringpop.AddListener(r)
+
+	labels, err := r.ringpop.Labels()
+	if err != nil {
+		r.logger.Fatal("unable to get ring pop labels", tag.Error(err))
+	}
+
+	if err = labels.Set(roleKey, r.service); err != nil {
+		r.logger.Fatal("unable to set ring pop labels", tag.Error(err))
+	}
+}
+
+// HandleEvent handles updates from ringpop
+func (r *Provider) HandleEvent(
+	event events.Event,
+) {
+	// We only care about RingChangedEvent
+	e, ok := event.(events.RingChangedEvent)
+	if !ok {
+		return
+	}
+
+	r.logger.Info("Received a ringpop ring changed event")
+	// Marshall the event object into the required type
+	change := &membership.ChangedEvent{
+		HostsAdded:   e.ServersAdded,
+		HostsUpdated: e.ServersUpdated,
+		HostsRemoved: e.ServersRemoved,
+	}
+
+	// Notify subscribers
+	r.smu.RLock()
+	defer r.smu.RUnlock()
+
+	for name, ch := range r.subscribers {
+		select {
+		case ch <- change:
+		default:
+			r.logger.Error("Failed to send listener notification, channel full", tag.Subscriber(name))
+		}
+	}
+
+}
+
+func (r *Provider) SelfEvict() error {
+	return r.ringpop.SelfEvict()
+}
+
+// GetMembers returns all hosts with a specified role value
+func (r *Provider) GetMembers(service string) ([]string, error) {
+	return r.ringpop.GetReachableMembers(swim.MemberWithLabelAndValue(roleKey, service))
+}
+
+// WhoAmI returns address of this instance
+func (r *Provider) WhoAmI() (string, error) {
+	return r.ringpop.WhoAmI()
+}
+
+// Stop stops ringpop
+func (r *Provider) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&r.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	r.ringpop.Destroy()
+}
+
+// Subscribe allows to be subscribed for ring changes
+func (r *Provider) Subscribe(name string, notifyChannel chan<- *membership.ChangedEvent) error {
+	r.smu.RLock()
+	defer r.smu.RUnlock()
```

Comment:
```suggestion
	r.smu.Lock()
	defer r.smu.Unlock()
```
This isn't a read operation, it can currently race with the iteration